### PR TITLE
Shims: make the `__swift_ssize_t` handling more portable

### PR DESCRIPTION
**Pull request by compnerd**
Unfortunately, `cl` does not support C11.  Guard the `_Generic` approach and in
C++ mode use `std::make_signed`.  Thanks to Jordan Rose for the reminder about
the type_traits helper!

Replace this paragraph with a description of your changes and rationale. Provide links to external references/discussions if appropriate.

Resolves [SR-NNNN](https://bugs.swift.org/browse/SR-NNNN).

### CODE REVIEW

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
CC: @jrose-apple @jckarter @mikeash

**Comment by mikeash** [comment]
Looks reasonable. Might be good to add a `#else #error "Don't know how to make __swift_ssize_t with this compiler"` or something along those lines.

File: stdlib/public/SwiftShims/SwiftStddef.h [hunk 1]
```diff
@@ -29,11 +29,16 @@ typedef __SIZE_TYPE__ __swift_size_t;
...
+#if __STDC_VERSION__-0 >= 201112l
 typedef __typeof__(_Generic((__swift_size_t)0,                                 \
                             unsigned long long int : (long long int)0,         \
                             unsigned long int : (long int)0,                   \
                             unsigned int : (int)0,                             \
                             unsigned short : (short)0,                         \
                             unsigned char : (signed char)0)) __swift_ssize_t;
+#elif defined(__cplusplus)
+#include <type_traits>
+using __swift_ssize_t = std::make_signed<std::size_t>::type;
```
Comment (jrose-apple) [inline stdlib/public/SwiftShims/SwiftStddef.h]: Nitpick: this should use `__swift_size_t`, to not implicitly depend on `<cstddef>`.

File: stdlib/public/SwiftShims/SwiftStddef.h [hunk 2]
```diff
@@ -29,11 +29,16 @@ typedef __SIZE_TYPE__ __swift_size_t;
...
+#endif
```
Comment (jrose-apple) [inline stdlib/public/SwiftShims/SwiftStddef.h]: Can you include an `else` branch here, so that we get a clear error instead of just not defining the type?

**Review by jrose-apple — APPROVED** [review]

File: stdlib/public/SwiftShims/SwiftStddef.h (hunk 2 above)
Comment (compnerd) [reply stdlib/public/SwiftShims/SwiftStddef.h]: I like it 👍

File: stdlib/public/SwiftShims/SwiftStddef.h (hunk 1 above)
Comment (compnerd) [reply stdlib/public/SwiftShims/SwiftStddef.h]: Good idea!

**Comment by compnerd** [comment]
@swift-ci please test and merge

**Comment by compnerd** [comment]
@swift-ci please test and merge

**Pull request merged**